# CORAC [![License: MIT](https://img.shields.io/badge/License-MIT-yellow.svg)](https://github.com/gamazonlab/CORAC/blob/master/LICENSE) 

## A phenome-wide scan reveals convergence of common and rare variant associations
Dan Zhou, Yuan Zhou, Yue Xu, Ran Meng, Eric R. Gamazon.  
<i>Genome Medicine</i> in press.

## For questions:  
Dan Zhou <zdangm@gmail.com>  
Eric R. Gamazon <ericgamazon@gmail.com>  

### Background 
Common and rare variants contribute to the etiology of complex traits. However, the extent to which the phenotypic effects of common and rare variants involve shared molecular mediators remains poorly understood. The question is essential to the basic and translational goals of the science of genomics, with critical basic-science, methodological, and clinical consequences. 

### Methods 
Leveraging the latest release of whole-exome sequencing (WES, for rare variants) and genome-wide association study (GWAS, for common variants) data from the UK Biobank, we developed a metric, the COmmon variant and RAre variant Convergence signature (CORAC) signature, to quantify the convergence for a broad range of complex traits. We characterized the relationship between CORAC and effective sample size across phenome-wide association studies. 

### Results 
We found that the signature is positively correlated with effective sample size (Spearman œÅ = 0.594, P < 2.2e-16), indicating increased functional convergence of trait-associated genetic variation, across the allele frequency spectrum, with increased power. Sensitivity analyses, including accounting for heteroskedasticity and varying the number of detected association signals, further strengthened the validity of the finding. In addition, consistent with empirical data, extensive simulations showed that negative selection, in line with enhancing polygenicity, has a dampening effect on the convergence signature. Methodologically, leveraging the convergence leads to enhanced association analysis. 

### Conclusions 
The presented framework for the convergence signature has important implications for fine-mapping strategies and drug discovery efforts. In addition, our study provides a blueprint for the expectation from future large-scale whole-genome sequencing (WGS)/WES and sheds methodological light on post-GWAS studies.

